rclcpp/test/rclcpp/executors/test_events_executor.cpp

```diff
@@ -479,14 +479,21 @@ TEST_F(TestEventsExecutor, test_default_incompatible_qos_callbacks)
   const auto timeout = std::chrono::seconds(10);
   ex.spin_until_future_complete(log_msgs_future, timeout);
 
-  EXPECT_EQ(
-    "New subscription discovered on topic '/test_topic', requesting incompatible QoS. "
-    "No messages will be sent to it. Last incompatible policy: DURABILITY_QOS_POLICY",
-    pub_log_msg);
-  EXPECT_EQ(
-    "New publisher discovered on topic '/test_topic', offering incompatible QoS. "
-    "No messages will be sent to it. Last incompatible policy: DURABILITY_QOS_POLICY",
-    sub_log_msg);
+
+  std::string rmw_implementation_str = std::string(rmw_get_implementation_identifier());
+  if (rmw_implementation_str == "rmw_zenoh_cpp") {
+    EXPECT_EQ("", pub_log_msg);
+    EXPECT_EQ("", sub_log_msg);
+  } else {
+    EXPECT_EQ(
+      "New subscription discovered on topic '/test_topic', requesting incompatible QoS. "
+      "No messages will be sent to it. Last incompatible policy: DURABILITY_QOS_POLICY",
+      pub_log_msg);
+    EXPECT_EQ(
+      "New publisher discovered on topic '/test_topic', offering incompatible QoS. "
+      "No messages will be sent to it. Last incompatible policy: DURABILITY_QOS_POLICY",
+      sub_log_msg);
+  }
```

Comment:
I'm not a big fan of hard-coding `rmw_zenoh_cpp` in here, particularly since it is not yet in the core.

I think we can do this better by using the QoS compatibility checker: https://github.com/ros2/rclcpp/blob/0be8aa013e4ed94bcab88527887e85130816d256/rclcpp/src/rclcpp/qos.cpp#L349 .  This calls down into the RMW to see if the QoS are compatible.

Thus, I think something like the following would work:

```
  rclcpp::QoSCheckCompatibleResult qos_compatible = rclcpp::qos_check_compatible(publisher->get_actual_qos(), subscription->get_actual_qos());
  if (qos_compatible.compatibility == rclcpp::QoScompatibility::Error) {
    EXPECT_EQ(
      "New subscription discovered on topic '/test_topic', requesting incompatible QoS. "
      "No messages will be sent to it. Last incompatible policy: DURABILITY_QOS_POLICY",
      pub_log_msg);
    EXPECT_EQ(
      "New publisher discovered on topic '/test_topic', offering incompatible QoS. "
      "No messages will be sent to it. Last incompatible policy: DURABILITY_QOS_POLICY",
      sub_log_msg);
  } else {
    EXPECT_EQ("", pub_log_msg);
    EXPECT_EQ("", sub_log_msg);
  }
```

(this is similar to what I did in https://github.com/ros2/rclcpp/pull/2653.